src/core/qgspallabeling.cpp

```diff
@@ -1550,8 +1554,38 @@ void QgsPalLayerSettings::calculateLabelSize( const QFontMetricsF *fm, const QSt
       {
         maxLineLength = std::max( maxLineLength, line.length() );
       }
-      h = labelHeight * maxLineLength + ( maxLineLength - 1 ) * letterSpacing;
+      h = fm->ascent() * maxLineLength + ( maxLineLength - 1 ) * letterSpacing;
       break;
+    }
+
+    case QgsTextFormat::RotationBasedOrientation:
+    {
+      double widthHorizontal = 0.0;
```

Comment:
If I'm reading this right - it's returning the largest size possible, right? I.e. not the actual label size, but the worst case scenario?

If so, that's far from ideal, because it means candidates in the unused areas will be stripped, labels won't be placed near the edges of the map, polygons set to place only labels which fit inside won't be labeled, etc. 

I think you're going to have to push this back into the pal library -- so PAL would need a concept of horizontal or vertical label size, and it would use the appropriate one based on the candidate's rotation.